tests/e2e/gc_rwx_statefulsets.go

```diff
@@ -0,0 +1,561 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/onsi/ginkgo"
+	"github.com/onsi/gomega"
+	cnstypes "github.com/vmware/govmomi/cns/types"
+	apps "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/kubernetes/test/e2e/framework"
+	fnodes "k8s.io/kubernetes/test/e2e/framework/node"
+	fpod "k8s.io/kubernetes/test/e2e/framework/pod"
+	fss "k8s.io/kubernetes/test/e2e/framework/statefulset"
+)
+
+var _ = ginkgo.Describe("[rwm-csi-tkg] File Volume Provision with Statefulsets", func() {
+	f := framework.NewDefaultFramework("rwx-tkg-sts")
+	var (
+		client            clientset.Interface
+		namespace         string
+		scParameters      map[string]string
+		storagePolicyName string
+		volHealthCheck    bool
+		isSTSDeleted      bool
+		isServiceDeleted  bool
+	)
+	ginkgo.BeforeEach(func() {
+		client = f.ClientSet
+		// TODO: Read value from command line
+		volHealthCheck = true
+		isSTSDeleted = false
+		isServiceDeleted = false
+		namespace = getNamespaceToRunTests(f)
+		scParameters = make(map[string]string)
+		storagePolicyName = GetAndExpectStringEnvVar(envStoragePolicyNameForSharedDatastores)
+		bootstrap()
+		nodeList, err := fnodes.GetReadySchedulableNodes(f.ClientSet)
+		framework.ExpectNoError(err, "Unable to find ready and schedulable Node")
+		if !(len(nodeList.Items) > 0) {
+			framework.Failf("Unable to find ready and schedulable Node")
+		}
+	})
+
+	ginkgo.AfterEach(func() {
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+		if !isSTSDeleted {
+			ginkgo.By(fmt.Sprintf("Deleting all statefulsets in namespace: %v", namespace))
+			fss.DeleteAllStatefulSets(client, namespace)
+		}
+
+		if !isServiceDeleted {
+			ginkgo.By(fmt.Sprintf("Deleting service nginx in namespace: %v", namespace))
+			err := client.CoreV1().Services(namespace).Delete(ctx, servicename, *metav1.NewDeleteOptions(0))
+			gomega.Expect(err).NotTo(gomega.HaveOccurred())
+		}
+	})
+
+	/*
+		Test to verify file volume provision with statefulsets
+		1. Create Storage class
+		2. Create Nginx service
+		3. Create Nginx statefulset with 3 replicas and
+				podManagementPolicy=OrderedReady using the Storage
+				Policy obtained in Step 1 (each replica to use a dedicated file volume)
+		4. Wait until all Pods are ready and PVCs are bounded with PV
+		5. Verify CnsVolumeMetadata CRD are created
+		6. Verify CnsFileAccessConfig CRD are created
+		7. Verify health status of PVC
+		8. Verify volume is created on CNS by using CNSQuery API and also check metadata is pushed to CNS
+		9. Scale down statefulset to 2 replicas
+		10. Scale-up statefulset to 3 replicas
+		11. Scale down statefulset to 0 replicas and delete all pods.
+		12. Delete the statefulset
```

Comment:
any reason why you reduce the sts replica count to 0, why not directly delete the sts?

Reply:
We are verifying the scale-up and scale-down operations works or not here.